### PR TITLE
[ROCm][Bugfix][MLA] Fix FP8 KV cache for sparse MLA decode on gfx950 (DSv3.2)

### PR DESCRIPTION
**Pull request by maeehart**
## Summary

Fixes empty / garbage output from the ROCm sparse MLA backend when running DSv3.2 with `--kv-cache-dtype=fp8` on gfx950 (MI35x, TP=4). On this configuration the AITER decode kernel has no working `q.fp8 + kv.fp8 + max_seqlen_q == 1` path for `nhead == 32`; the python wrapper's simulation that folds `nhead == 32 → 16` returns empty output, and the `fp8_attention` block added in #41217 silently routed into it. Pad Q's heads `32 → 64` so the kernel takes its native `gqa_ratio == 64` path, then strip the duplicate output rows. MQA-style attention is per-head independent so this is exact.

Also fixes a related FP8 K-scale calibration bug in `MLAAttention.calc_kv_scales` that affected MLA paths when `--calculate-kv-scales` was set: `_k_scale` was derived from `|kv_c_normed|.max()` only, but `concat_and_cache_mla` writes both `kv_c_normed` and `k_pe` into the FP8 cache with the same scale. `k_pe` (RoPE) lies in roughly `[-1, 1]` while `kv_c_normed` is `< 0.1` after RMSNorm, so the old calibration saturated `k_pe` to ±FP8 max. Take `max(|kv_c_normed|, |k_pe|)` instead.

## Validation

Hardware: 4 × MI355X (gfx950), TP=4, image `amdsiloai/vllm-private:nightly_SHA0fa8884_42532_42838_42864_39177_41675_aiterv0.1.14-rc0_aiterPR3075`, DSv3.2, sparse MLA backend (`ROCM_AITER_MLA_SPARSE`), v1 engine, `--max-model-len 4096`.

`lm_eval --tasks gsm8k --num_fewshot 5`, full 1319 samples, two runs each:

| Config | flexible-extract | strict-match |
|---|---|---|
| bf16 KV (baseline) — run 1 / 2 | 0.8415 / 0.8453 | 0.8453 / 0.8506 |
| **fp8 KV with this PR + `--calculate-kv-scales`** — run 1 / 2 | **0.9507 / 0.9500** | **0.9515 / 0.9507** |

Without this PR, FP8 KV produces empty / garbage output and is unusable. With this PR, FP8 KV is functional and reaches ~95% on GSM8K — actually higher than the bf16 baseline on this image, because the bf16 sparse path also routes through the same broken `nhead == 32 → 16` simulation that we're bypassing. Fixing bf16 sparse separately is out of scope here.

`vllm bench serve` random workload, ignore-eos:

| Config | mc=8 in=1000 out=100, out tok/s | mc=32 in=2000 out=1000, out tok/s |
|---|---|---|
| bf16 KV (baseline) | 318.12 | 810.32 |
| fp8 KV with this PR | 301.80 (-5%) | 779.56 (-4%) |

The ~4% drop comes from the pad-to-64 doubling per-step attention compute. At this workload the decode step is dominated by MoE expert loads, not KV reads, so FP8's halved KV bandwidth doesn't pay off here. The actual production benefit of FP8 KV is the half-size cache in HBM (more concurrent requests / longer context fit at the same memory budget); a future native AITER `nhead == 32 + max_seqlen_q == 1 + fp8` kernel on gfx950 (or wiring up a `nhead == 16 × 2` split using the existing native `nhead == 16` path) would eliminate the 2× compute waste and let FP8 KV beat bf16 KV on raw throughput too. Tracked as the `TODO` in `_needs_fp8_pad_to_64_for_aiter_native`.

## Scope and risk

- ROCm sparse MLA backend file `vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py` and `vllm/model_executor/layers/attention/mla_attention.py`.
- Pad-to-64 path is gated to `current_platform.is_rocm() AND device_name contains "mi35" AND num_heads == 32`; all other configurations follow exactly the same code path as before.
- The `_k_scale` calibration change only activates with `--calculate-kv-scales`; default behavior is unchanged.
- Not tested on MI300X (gfx942) FP8 KV with sparse MLA — that path stays unchanged.

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: vllm/model_executor/layers/attention/mla_attention.py [hunk 1]
```diff
@@ -944,9 +944,13 @@ def calc_kv_scales(
         k_range = getattr(self, "k_range", torch.tensor(1.0))
         v_range = getattr(self, "v_range", torch.tensor(1.0))
 
-        self._q_scale.copy_(torch.abs(q).max() / q_range)
-        # kv_c_normed is the compressed KV representation; use it for k/v
-        kv_abs_max = torch.abs(kv_c_normed).max()
+        self._q_scale.copy_(q.abs().amax() / q_range)
```
Comment (Rohan138) [inline vllm/model_executor/layers/attention/mla_attention.py:947]: Can we move this into a separate PR? This probably needs additional review for NV MLA backends.

File: vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py [hunk 2]
```diff
@@ -421,6 +442,15 @@ def __init__(
             kv_cache_dtype_str = "bf16"
         kv_dtype = dtypes.d_dtypes.get(kv_cache_dtype_str, dtypes.bf16)
 
+        # Match what forward_mqa will pass to the kernel when the gfx950
```
Comment (Rohan138) [inline vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py:445]: Can we explicitly mention https://github.com/ROCm/aiter/pull/3233 in the comments? (I presume that PR will fix the error once we bump AITER to a newer commit)

Also, could we try consolidating this with the `get_mla_padded_q`/`get_mla_unpadded_q` logic in `AiterMLAHelper`: https://github.com/vllm-project/vllm/blob/main/vllm/v1/attention/backends/mla/rocm_aiter_mla.py#L392? Instead of adding separate logic for this here? Especially since we'll need to come back and remove this once we bump AITER

File: vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py [hunk 3]
```diff
@@ -342,6 +342,27 @@ class ROCMAiterMLASparseMetadata(AttentionMetadata):
...
+    natively supported.  ``nhead==32`` (DSv3.2 TP=4 per-rank) is not:
+    the python wrapper folds it to ``nhead==16`` which lacks a working
+    sparse fp8 kernel on this arch and silently produces empty output.
+
+    Padding ``nhead 32 -> 64`` (Q-head replication; MQA is per-head
+    independent so this is exact) routes to the native ``gqa_ratio==64``
+    kernel.  TODO: remove once AITER ships a native sparse fp8 kernel
+    for ``nhead==32``/``max_seqlen_q==1`` on gfx950.
+    """
+    if not current_platform.is_rocm():
```
Comment (Rohan138) [inline vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py:360]: We don't need this helper function, can directly check `on_gfx950() and num_heads == 32` https://github.com/vllm-project/vllm/blob/main/vllm/platforms/rocm.py#L291

**Comment by maeehart** [comment]
I need to see if this can be fixed in the AITER side. I changed this to draft.

File: vllm/model_executor/layers/attention/mla_attention.py [hunk 4]
```diff
@@ -945,8 +945,18 @@ def calc_kv_scales(
...
+        # the FP8 cache with the same ``_k_scale``.  ``k_pe`` typically
+        # has substantially larger magnitudes than ``kv_c_normed``
+        # (RoPE values lie in roughly ``[-1, 1]`` while ``kv_c_normed``
+        # is often <0.1 after RMSNorm), so calibrating ``_k_scale``
+        # from only ``kv_c_normed`` saturates ``k_pe`` to FP8 max and
+        # destroys positional information, producing garbage outputs
+        # in MLA decode/sparse-decode.  Take the max over both halves
+        # to give ``k_pe`` enough headroom while keeping ``kv_c_normed``
+        # representable.
+        kv_abs_max = torch.maximum(torch.abs(kv_c_normed).max(), torch.abs(k_pe).max())
```
[the remark this answers is not in the document]
Comment (maeehart) [reply vllm/model_executor/layers/attention/mla_attention.py]: Applied in `210ef752` at the time. #49389 has since removed the runtime KV-scale calibration path, so this change will not be carried forward.

File: vllm/model_executor/layers/attention/mla_attention.py (hunk 1 above)
Comment (maeehart) [reply vllm/model_executor/layers/attention/mla_attention.py:947]: Agreed. This should have been a separate PR. #49389 has since removed runtime calibration entirely, so there is no calibration change left to split out or review for NVIDIA.

**Comment by maeehart** [comment]
Closing this draft as obsolete. #46997 fixed the gfx950 FP8 head-fold failure at the metadata root, so the padding workaround is no longer needed. Also, #49389 removed `--calculate-kv-scales` and `MLAAttention.calc_kv_scales`, so the calibration commit no longer applies.
